File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/validation/DateRangeValidator.java
package com.github.arhor.simple.expense.tracker.validation;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

import com.github.arhor.simple.expense.tracker.web.api.DateRangeCriteria;

public class DateRangeValidator implements ConstraintValidator<ValidDateRange, DateRangeCriteria> {

    @Override
    public boolean isValid(final DateRangeCriteria value, final ConstraintValidatorContext context) {
        if (value == null || value.startDate() == null || value.endDate() == null) {
            return true;
        }
        return value.startDate().isAfter(value.endDate());
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/validation/DateRangeValidator.kt
package com.github.arhor.simple.expense.tracker.validation

import com.github.arhor.simple.expense.tracker.DateRangeCriteria
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class DateRangeValidator : ConstraintValidator<ValidDateRange, DateRangeCriteria> {

    override fun isValid(value: DateRangeCriteria?, context: ConstraintValidatorContext?): Boolean {
        value?.let { (startDate, endDate) ->
            if ((startDate != null) && (endDate != null)) {
                return startDate.isBefore(endDate)
                    || startDate.isEqual(endDate)
            }
        }
        return true
    }
}
